### PR TITLE
Feature/encoding

### PR DESCRIPTION
**Pull request by lmtan91**

### CODE REVIEW

File: pkg/router.go [hunk 1]
```diff
@@ -159,11 +161,48 @@ func PointWrite(m *nmodule.Module, r *router.Request) ([]byte, error) {
...
+			} else if pnt.Name == "vertical position" {
+				endec.EncodeData(serialData, uint8(floatValue), endec.MDK_UINT_8, 6)
+			} else if pnt.Name == "horizontal position" {
+
+			}
+		}
+	}
 
+	// TODO: Encrypt the encoded data by using aesutils.Encrypt method
+	// aesutils.Encrypt()
```
Comment (Shiny380) [inline pkg/router.go:197]: uncomment here to encrypt and add loraraw data

File: pkg/router.go [hunk 2]
```diff
@@ -159,11 +161,48 @@ func PointWrite(m *nmodule.Module, r *router.Request) ([]byte, error) {
...
+				endec.EncodeData(serialData, floatValue, endec.MDK_TEMP, 4)
+			} else if pnt.Name == "fan speed" {
+				endec.EncodeData(serialData, uint8(floatValue), endec.MDK_UINT_8, 5)
+			} else if pnt.Name == "vertical position" {
+				endec.EncodeData(serialData, uint8(floatValue), endec.MDK_UINT_8, 6)
+			} else if pnt.Name == "horizontal position" {
+
+			}
+		}
+	}
```
Comment (Shiny380) [inline pkg/router.go:194]: @maharjanaman can you please show tan how to get the address id and data type from the point model as per these https://github.com/NubeIO/module-core-loraraw/pull/39/files#diff-0a2c23f05a4f5c27ab49210097aac86bd8f07549b36a4a81ac069ee9cf74f449R16-R25

**Review by Shiny380 — CHANGES_REQUESTED** [review]

File: endec/rubix.go [hunk 3]
```diff
@@ -318,43 +221,25 @@ func decodeData(serialData *SerialData, header MetaDataKey, data interface{}) er
...
+		return "Cool Temperature"
+	} else if *pos == 4 {
+		return "Heat Temperature"
+	} else if *pos == 5 {
+		return "Fan Speed"
+	} else if *pos == 6 {
+		return "Vertical Position"
+	} else if *pos == 7 {
+		return "Horizontal Position"
+	}
```
Comment (Shiny380) [inline endec/rubix.go:242]: @lmtan91 just to be clear, we don't want any code in here related to the FGA OEM product. This Rubix-Data-Encoding is meant to be generic for any device that implements it. Don't forget that the Dorma uart-to-lora oem product also uses this encoding format so their point IDs will mean something completely different. Therefore it is up to the user to name these points correctly from the UI (CE). This block should be removed

File: pkg/router.go (hunk 2 above)
Comment (Shiny380) [reply pkg/router.go:194]: @lmtan91 this is the same as the other comment re this needing to be generic/agnostic

File: schema/device.go [hunk 4]
```diff
@@ -12,6 +12,13 @@ const (
 	DeviceModelRubix        = "Rubix"
 )
 
+type DeviceKey struct {
+	Type  string `json:"type" default:"string"`
+	Title string `json:"title" default:"Device Key"`
+	Min   int    `json:"minLength" default:"32"`
+	Max   int    `json:"maxLength" default:"32"`
```
Comment (Shiny380) [inline schema/device.go:19]: @maharjanaman same thing re the min/max spelling. please replace with the correct spelling

**Review by Shiny380 — CHANGES_REQUESTED** [review]

File: schema/device.go (hunk 4 above)
Comment (maharjanaman) [reply schema/device.go:19]: @Shiny380 Got it. Will make the changes. But other modules is using the same https://github.com/NubeIO/module-core-lorawan/blob/0b6ba552f1b79e89fd5d13de86aa43fd5eac28f7/schema/device.go#L43 and some struct in loraraw is still using same convention. May be something to refactor later.

Comment (maharjanaman) [reply schema/device.go:19]: Json property, `minLength` and `maxLength` is the same. I checked with UI and it's working fine.

Comment (Shiny380) [reply schema/device.go:19]: ok thanks!
I just knew that `max`/`min` didn't work (which we made a few mistakes in the past) and `maximum`/`minimum` did, interesting that this also works

Comment (Shiny380) [reply schema/device.go:19]: if it works then no need to change!

File: pkg/router.go [hunk 5]
```diff
@@ -158,12 +163,69 @@ func PointWrite(m *nmodule.Module, r *router.Request) ([]byte, error) {
...
+				endec.EncodeData(serialData, int8(floatValue), endec.MetaDataKey(pointDataType), uint8(*pnt.AddressID))
+			} else if endec.MetaDataKey(pointDataType) == endec.MDK_INT_16 {
+				endec.EncodeData(serialData, int16(floatValue), endec.MetaDataKey(pointDataType), uint8(*pnt.AddressID))
+			} else if endec.MetaDataKey(pointDataType) == endec.MDK_INT_32 {
+				endec.EncodeData(serialData, int32(floatValue), endec.MetaDataKey(pointDataType), uint8(*pnt.AddressID))
+			} else if endec.MetaDataKey(pointDataType) == endec.MDK_INT_64 {
+				endec.EncodeData(serialData, int64(floatValue), endec.MetaDataKey(pointDataType), uint8(*pnt.AddressID))
+			} else {
+				endec.EncodeData(serialData, floatValue, endec.MetaDataKey(pointDataType), uint8(*pnt.AddressID))
+			}
```
Comment (Shiny380) [inline pkg/router.go:202]: be careful de-referencing `*pnt.AddressID`
@maharjanaman can you please add that util function to this. if nil, return error instead maybe

**Review by Shiny380 — CHANGES_REQUESTED** [review]

**Review by Shiny380 — APPROVED** [review]

**Pull request merged**